### PR TITLE
feature: add `apply_conf_change` method for `RawCurp`

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
#306 
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
apply config change to curp
* what changes does this pull request make?
refactor data structures via `DashMap`, and modify them when apply_conf_change
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: curp/src/server/curp_node.rs [hunk 1]
```diff
@@ -320,22 +329,18 @@ impl<C: 'static + Command, RC: RoleChange + 'static> CurpNode<C, RC> {
...
-                Arc::clone(&connect),
-                Arc::clone(&sync_event),
-            )
-            .await;
+            let peers = curp.cluster().peers_addrs();
+            let futs = rpc::connect(peers).await.map(|(id, connect)| {
+                let sync_event = curp.sync_event(id);
+                Self::sync_follower_task(Arc::clone(&curp), connect, sync_event)
+            });
+            _ = futures::future::join_all(futs).await;
```
Comment (bsbds) [inline curp/src/server/curp_node.rs:343]: Could spawn a tokio task for each `sync_follower_task` achieve better performance? As they could run on different thread.

File: curp/src/server/raw_curp/mod.rs [hunk 2]
```diff
@@ -791,9 +775,104 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
...
+    #[allow(clippy::unimplemented)] // TODO: remove
+    fn switch_config(&self, conf_change: ConfChange) -> bool {
+        let node_id = conf_change.node_id;
+        let conf_change_type =
+            ConfChangeType::from_i32(conf_change.change_type).unwrap_or_else(|| {
+                unreachable!("conf change type {} should valid", conf_change.change_type)
+            });
+        match conf_change_type {
+            ConfChangeType::Add => {
+                let member = Member::new(node_id, "", conf_change.address);
```
Comment (bsbds) [inline curp/src/server/raw_curp/mod.rs:851]: The member name is empty here, should the ConfChange contains a member name?

Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:851]: The `MemberAddRequest` of `etcd` does not contain `name`. add `name` here will break compatibility.

File: curp/src/server/curp_node.rs (hunk 1 above)
Comment (themanforfree) [reply curp/src/server/curp_node.rs:347]: Fixed

File: curp/src/server/raw_curp/state.rs [hunk 3]
```diff
@@ -48,19 +51,29 @@ pub(super) struct CandidateState<C> {
...
+pub(super) struct FollowerStatus {
     /// Index of the next log entry to send to that follower
     next_index: LogIndex,
     /// Index of highest log entry known to be replicated on that follower
     match_index: LogIndex,
 }
 
+impl FollowerStatus {
+    /// Create a `FollowerStatus`
+    pub(super) fn new(next_index: LogIndex, match_index: LogIndex) -> Self {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/server/raw_curp/state.rs]: Fixed

File: curp/src/server/raw_curp/state.rs [hunk 4]
```diff
@@ -103,25 +116,41 @@ impl LeaderState {
...
             .get(&id)
             .unwrap_or_else(|| unreachable!("no status for {id}"))
-            .lock()
+    }
+
+    /// Get mutable status for a server
+    fn get_status_mut(&self, id: ServerId) -> RefMut<'_, u64, FollowerStatus> {
+        self.statuses
+            .get_mut(&id)
+            .unwrap_or_else(|| unreachable!("no status for {id}"))
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/server/raw_curp/state.rs:152]: Normal operation does not attempt to get the status of a non-existent follower

File: curp/src/server/raw_curp/tests.rs [hunk 5]
```diff
@@ -670,3 +677,105 @@ fn is_synced_should_return_true_when_followers_caught_up_with_leader() {
     curp.lst.update_match_index(s2_id, 3);
     assert!(curp.is_synced());
 }
+
+#[traced_test]
+#[test]
+fn add_node() {
```
Comment (Phoenix500526) [inline curp/src/server/raw_curp/tests.rs:683]: Please add your expectations in the test name like other tests do.

File: curp/src/server/raw_curp/tests.rs [hunk 6]
```diff
@@ -29,6 +29,13 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
         self.st.read().role
     }
 
+    fn contains(&self, id: ServerId) -> bool {
```
Comment (Phoenix500526) [inline curp/src/server/raw_curp/tests.rs:32]: What is the semantics of this method? To check whether the given node is the current member of this cluster? If so, why can not directly check via `cluster.get_members`?

Comment (themanforfree) [reply curp/src/server/raw_curp/tests.rs:32]: This is a test utility to confirm that the node modifies all the parts that need to be modified.

File: curp/src/server/raw_curp/tests.rs (hunk 5 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/tests.rs:683]: Fixed

File: curp/src/server/curp_node.rs [hunk 7]
```diff
@@ -120,7 +121,7 @@ pub(super) struct CurpNode<C: Command, RC: RoleChange> {
 
 // handlers
 impl<C: 'static + Command, RC: RoleChange + 'static> CurpNode<C, RC> {
-    /// Handle "propose" requests
+    /// Handle `Propose` requests
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/server/curp_node.rs:124]: Fixed

File: curp/src/members.rs [hunk 8]
```diff
@@ -56,47 +77,85 @@ impl ClusterInfo {
...
+    }
+
     /// Get server address via server id
     #[must_use]
     #[inline]
-    pub fn address(&self, id: ServerId) -> Option<&str> {
+    pub fn address(&self, id: ServerId) -> Option<String> {
         self.members
-            .values()
+            .iter()
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/members.rs:116]: FIxed

File: curp/src/members.rs [hunk 9]
```diff
@@ -56,47 +77,85 @@ impl ClusterInfo {
         cluster_info
     }
 
+    /// Get all members
+    #[must_use]
+    #[inline]
+    pub fn get_members(&self) -> HashMap<ServerId, Member> {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/members.rs:83]: Renamed

File: curp/proto/message.proto [hunk 10]
```diff
@@ -89,11 +89,35 @@ message ShutdownResponse {
...
+        Remove = 2;
+        Update = 3;
+    }
+    message ConfChange {
+        ConfChangeType change_type = 1;
+        uint64 node_id = 2;
+        string address = 3;
+    }
+    string id = 1;
+    repeated ConfChange changes = 2;
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/proto/message.proto:107]: we won't have several `ConfChange`  in the current implementation. Here is to prepare for joint consensus.

Comment (iGxnon) [inline curp/proto/message.proto:107]: When a node can have multiple addresses, does this mean using multiple ConfChange for the same node_id?

File: curp/src/server/raw_curp/mod.rs [hunk 11]
```diff
@@ -866,6 +851,100 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
...
+            ConfChangeType::Update => {
+                if statuses_ids.get(&node_id).is_none() || !config.contains(node_id) {
+                    return Err(ApplyConfChangeError::NodeNotExists(node_id));
+                }
+            }
+            ConfChangeType::AddLearner => {
+                unimplemented!("learner node is not supported yet");
+            }
+        }
+        if statuses_ids.len() < 3 || config.voters() != &statuses_ids {
```
Comment (iGxnon) [inline curp/src/server/raw_curp/mod.rs:907]: Is it possible for us to reduce a 3-node cluster to just one node? Will there be an error here?

File: curp/src/server/curp_node.rs [hunk 12]
```diff
@@ -161,6 +162,15 @@ impl<C: 'static + Command, RC: RoleChange + 'static> CurpNode<C, RC> {
         Ok(resp)
     }
 
+    /// Handle `ProposeConfChange` requests
+    #[allow(clippy::todo, clippy::unused_async)] // TODO: implement this
+    pub(super) async fn propose_conf_change(
+        &self,
+        _req: ProposeConfChangeRequest,
+    ) -> Result<ProposeConfChangeResponse, CurpError> {
+        todo!("propose_conf_change")
```
Comment (iGxnon) [inline curp/src/server/curp_node.rs:171]: Can't `raw_curp::apply_conf_change` be used here yet?

File: curp/src/server/cmd_worker/mod.rs [hunk 13]
```diff
@@ -246,7 +246,7 @@ struct TaskRx<C: Command>(flume::Receiver<Task<C>>);
 
 /// Send cmd to background execution worker
 #[cfg_attr(test, automock)]
-pub(super) trait CEEventTxApi<C: Command + 'static>: Send + Sync + 'static {
+pub(super) trait CEEventTxApi<C: Command + 'static>: Send + Sync + 'static + Debug {
```
Comment (iGxnon) [inline curp/src/server/cmd_worker/mod.rs:249]: What is the role of the supertrait `Debug` here?

**Comment by iGxnon** [comment]
BTW, can we directly use the etcd's protobuf definition of ConfChange? This way, we won't need to add an extra layer of translation in xline to the current definition of ConfChange.

File: curp/src/server/curp_node.rs (hunk 12 above)
Comment (themanforfree) [reply curp/src/server/curp_node.rs:171]: Similar to `etcd`, `apply_conf_change` will be called when this log entry is apply.

File: curp/src/server/raw_curp/mod.rs (hunk 11 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:907]: Removing a node from a 3-node cluster will be rejected

File: curp/proto/message.proto (hunk 10 above)
Comment (themanforfree) [reply curp/proto/message.proto:107]: No, we need no add `repeated` to address if  a node can have multiple addresse. 
Added

File: curp/src/server/raw_curp/mod.rs (hunk 11 above)
Comment (iGxnon) [reply curp/src/server/raw_curp/mod.rs:907]: Cannot turn the cluster into a single node? What about etcd?

File: curp/src/server/raw_curp/state.rs [hunk 14]
```diff
@@ -106,19 +115,29 @@ impl LeaderState {
...
-                            match_index: 0,
-                        },
-                    )
-                })
+                .map(|o| (*o, FollowerStatus::default()))
                 .collect(),
         }
     }
 
+    /// Get all status for a server
```
Comment (rogercloud) [inline curp/src/server/raw_curp/state.rs:123]: Get status for all servers

File: curp/src/server/raw_curp/mod.rs (hunk 11 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:924]: #447

File: curp/src/server/cmd_worker/mod.rs (hunk 13 above)
Comment (themanforfree) [reply curp/src/server/cmd_worker/mod.rs]: Removed this and manually implement `Debug` for `Context`

File: curp/src/rpc/mod.rs [hunk 15]
```diff
@@ -442,3 +445,43 @@ impl ShutdownRequest {
         Self { id: id.into() }
     }
 }
+
+#[allow(dead_code)] // TODO: remove
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/rpc/mod.rs]: Fixed

File: curp/src/rpc/mod.rs [hunk 16]
```diff
@@ -442,3 +445,43 @@ impl ShutdownRequest {
...
+#[allow(clippy::as_conversions)] // those conversions are safe
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/rpc/mod.rs]: Fixed

File: curp/src/server/raw_curp/mod.rs [hunk 17]
```diff
@@ -866,6 +851,103 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
         let leader_commit_index = log_r.commit_index;
         self.lst.check_all(|f| f.match_index == leader_commit_index)
     }
+
+    /// Apply conf changes and return true if self node is removed
+    #[allow(unused)] // TODO: remove
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs]: Fixed

File: curp/src/server/raw_curp/mod.rs [hunk 18]
```diff
@@ -866,6 +851,103 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
...
+            unreachable!("conf change is empty");
+        };
+
+        self.check_new_config(&conf_change)?;
+
+        Ok(self.switch_config(conf_change))
+    }
+
+    /// Check if the new config is valid
+    #[allow(clippy::unimplemented)] // TODO: remove
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs]: Fixed

File: curp/src/server/raw_curp/mod.rs [hunk 19]
```diff
@@ -866,6 +851,103 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
...
+            .copied()
+            .chain([self.id()])
+            .collect::<HashSet<_>>();
+        let mut config = self.cst.map_lock(|cst_l| cst_l.config.clone());
+        let conf_change_type =
+            ConfChangeType::from_i32(conf_change.change_type).unwrap_or_else(|| {
+                unreachable!("conf change type {} should valid", conf_change.change_type)
+            });
+        let node_id = conf_change.node_id;
+        match conf_change_type {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:904]: The modified data structure in this function is a clone and does not cause side effects

File: curp/src/server/raw_curp/state.rs (hunk 14 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/state.rs]: Fixed

File: curp/src/server/curp_node.rs [hunk 20]
```diff
@@ -161,6 +162,15 @@ impl<C: 'static + Command, RC: RoleChange + 'static> CurpNode<C, RC> {
         Ok(resp)
     }
 
+    /// Handle `ProposeConfChange` requests
+    #[allow(clippy::todo, clippy::unused_async)] // TODO: implement this
```
Comment (rogercloud) [inline curp/src/server/curp_node.rs:166]: Please link it to an issue.

**Review by iGxnon — COMMENTED** [review]
LGTM

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by iGxnon — APPROVED** [review]

**Pull request merged**